src/type.js

```diff
@@ -0,0 +1,199 @@
+ * @return {function} Object validator.
+ */
+function createObjectOfValidator(expectedType) {
+	return (stateValue, stateKey) => {
+		let success = true;
+
+		for (let key in stateValue) {
+			success = expectedType(stateValue[key], null, true) === null;
+		}
+
+		if (!success) {
+			return 'Expected object of one type';
+		}
+
+		return null;
+	}
+}
+
+/**
+ * Validates an objects shape.
+ * @param {!Object} shapeTypes An object containing type validators for each key.
+ * @return {function} Shape validator.
+ */
+function createShapeTypeValidator(shapeTypes) {
+	return (stateValue, stateKey) => {
+		let success = true;
+
+		for (let key in shapeTypes) {
+			const validator = shapeTypes[key];
+
+			if (!validator) {
+				continue;
+			}
+
+			const stateValueForKey = stateValue[key];
+
+			if (stateValueForKey) {
+				const success = validator(stateValueForKey, null, true) === null;
+
+				if (!success) {
+					return 'Expected object with a specific shape';
+				}
+			}
+		}
+
+		return null;
+	}
+}
+
+/**
+ * Validates a specific primitive type.
+ * @param {!string} expectedType Name of primitive type.
+ * @return {function} Type validator.
+ */
+function createPrimitiveTypeValidator(expectedType) {
+	return (stateValue, stateName) => {
+		const stateType = getStateType(stateValue);
+
+		if (stateType !== expectedType) {
+			return `Expected type ${expectedType}`;
+		}
+		return null;
+	}
+}
+
+/**
+ * Validates if any of the type validators are successfull.
+ * @param {!array} arrayOfTypeValidators Array of type validators.
+ * @return {string|function} Type validator or error.
+ */
+function createUnionTypeValidator(arrayOfTypeValidators) {
+	if (!Array.isArray(arrayOfTypeValidators)) {
+		return 'Expected an array';
+	}
+
+	return (stateValue, stateName) => {
+		for (var i = 0; i < arrayOfTypeValidators.length; i++) {
+			var validator = arrayOfTypeValidators[i];
+
+			if (validator(stateValue, stateName, true) == null) {
+				return null;
+			}
+		}
+
+		return 'Expected one of given types.';
+	}
+}
+
+/**
+ * Wrapper function that is used to determine if console.error should be thrown.
+ * @param {!function} validator Type validator.
+ * @return {function} This is the result function passed to the state validator.
+ */
+function emitWarning(validator) {
+	function returnFn(stateValue, stateName, disableWarning) {
+		const expectedError = validator(stateValue, stateName);
+
+		if (!disableWarning && expectedError) {
+			let componentName;
+			let parentComponent;
+			let parentComponentName;
+
+			if (this) {
+				componentName = this.constructor.name;
+				parentComponent = this.getRenderer ? this.getRenderer().lastParentComponent_ : null;
+				parentComponentName = parentComponent ? parentComponent.constructor.name : null;
+			}
+
+			console.error(
+				`Warning: Invalid state '${stateName}' of type '${getStateType(stateValue)}' ` +
+				`passed to '${componentName}'. ${expectedError}.  Check render method of ${parentComponentName}.`
```

Comment:
The message will be a bit weird when there's no parent component, which can happen for the root component as well as for classes extending from `State` directly, instead of through `Component`. Would be good to guarantee good messages for these cases.